rfd/0151-database-permission-management.md

```diff
@@ -0,0 +1,355 @@
+- Product: @xinding33 || @klizhentas
+- Security: @reedloden || @jentfoo
+
+## What
+
+Extends [automated db user provisioning](0113-automatic-database-users.md) with
+permission management capabilities.
+
+## Why
+
+Managing database-level permissions is a natural extension of current Teleport
+RBAC capabilities. The described model should also integrate seamlessly with
+TAG, providing a future-proof solution.
+
+Database administrators will be able to use Teleport for both user and
+permission management.
+
+## Details
+
+### UX
+
+A set of global "database object import rules" resources are defined for the
+Teleport cluster. At certain points in time, the database schema is read and
+passed through the import rules. The individual import rules may apply custom
+labels. An database object resource will only be created if there is at least
+one import rule that matched it.
+
+Only object attributes (standard ones like `protocol` or custom from the
+`attributes` field), which are sourced from the object spec, are subject to
+matching against the import rules. Any labels present on an object, either from
+an import rule or another source, are not subject to be matched by the import
+rule. The permissions matching is different in this regard.
+
+Example: the `sales-prod` import rule, which imports all tables from `sales`
+Postgres database in prod:
+
+```yaml
+kind: db_object_import_rule
+metadata:
+  name: sales-prod
+spec:
+  db_labels:
+    env: prod
+  mappings:
+    - object_match:
+        - database: sales
+          object_kind: table
+          protocol: postgres
+      add_labels:
+        env: prod
+        product: sales
+  priority: 10
+version: v1
+```
+
+Example database object, with applied labels:
+
+```yaml
+kind: db_object
+metadata:
+  labels:
+    env: prod
+    product: sales
+  name: sales_main
+spec:
+  attributes:
+    attr1: custom attr1 value
+    attr2: custom attr2 value
+  database: sales
+  name: sales_main
+  object_kind: table
+  protocol: postgres
+  schema: public
+  service_name: sales-prod-123
+version: v1
+```
+
+The permissions to particular objects are defined in a role using the new field
+`db_permissions`. The permission specifies the object kind (e.g. `table`),
+permission to grant/revoke (`SELECT`) as well as labels to be matched.
+
+The matching is performed against database object:
+
+- attributes: standard and custom, sourced from the object spec, provided by
+  database-specific schema import code
+- resource labels: aside from standard labels, these can be manipulated by the
+  import rules.
+
+If both attributes and labels provide the same non-empty key, the attributes
+take preference.
+
+Example role `db-dev`, which grants `SELECT` permission to some objects, and
+revokes `UPDATE` from all others.
+
+```yaml
+kind: role
+metadata:
+  name: db-dev
+spec:
+  allow:
+    db_permissions:
+      - match:
+          product: sales
+          protocol: postgres
+        object_kind: table
```

Comment:
If we're selecting by labels, why do we need an extra object kind filter? I think labels are flexible enough, no?

Reply:
It would work, but we have a higher risk of errors. The `object_kind` here serves to purposes:
1. Ensure the `object_kind` is matched. This forces the user to think exactly which object kind is targeted.
2. Fix the `object_kind` for a particular permission. This ties the permission used to object kind.

Reply:
As mentioned in another comment, I have dropped the object kind filter as it makes the UX better, even if there is some extra prep work needed:
> I went ahead and simplified the permissions matching part; now it is only done via labels, and not object attributes (name, schema, ...). You can still get the same effective behaviour if you add an import rule which propagates the attributes that you need (which is some work), but the positive part is that the implementation gets easier to write and understand; the latter part is a win in my book, as it makes the UX better. The roles are shorter, too.